### PR TITLE
Added Linux Workflow

### PR DESCRIPTION
**Pull request by DanilBaibak**
Added Linux Wheels workflow. The workflow will run in shadow mode without upload binaries.

### Changes

- Added Linux Wheels workflow
- Added Linux Conda workflow

### TODO

- [x] Add the `BUILD_S3` env variable. Depend on the PR https://github.com/pytorch/test-infra/pull/1269
- [x] Fixed smoke tests issue for Linux Conda - https://github.com/pytorch/test-infra/pull/1292
- [x] Add smoke test - https://github.com/pytorch/data/pull/923

### CODE REVIEW

File: .github/workflows/build_conda_linux.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+name: Build Linux Conda
+
+on:
+  pull_request:
+  push:
+    branches:
+      - nightly
```
Comment (ejguan) [inline .github/workflows/build_conda_linux.yml:7]: I have seen a nightly branch created and updated daily. Wondering if there is a way for us to manual trigger nightly release and all those build workflows?

**Review by ejguan — COMMENTED** [review]
It seems conda releases are broken because of missing `portalocker`. We do have a `meta.yml` to track those dependencies for conda. Is there way to provide it to the workflow?

File: .github/workflows/build_conda_linux.yml (hunk 1 above)
Comment (DanilBaibak) [reply .github/workflows/build_conda_linux.yml:7]: We have a separate Chronos job to trigger a nightly build https://fburl.com/chronos/udouc4ht So we can trigger it also manually.

Comment (ejguan) [reply .github/workflows/build_conda_linux.yml:7]: Ohhh, thank you!

**Comment by ejguan** [comment]
BTW, another comment regarding enabling `BUILD_S3=1`. I was only able to compile the AWSSDK plugin to support minimum `manylinux_2_17` with all the following manual installation steps (AWSSDK doesn't work well with dynamic linking).
https://github.com/pytorch/data/blob/9e217f7154c2ed96d95e9d9eb9152178a4e0c6a2/.github/workflows/_build_test_upload.yml#L111-L137

File: test/smoke_test/smoke_test.py [hunk 2]
```diff
@@ -26,5 +26,3 @@ def s3_test():
     options = parser.parse_args()
     if options.s3:
         s3_test()
-
-    main()
```
Comment (ejguan) [inline test/smoke_test/smoke_test.py:30]: Thank you! I have no idea what I was thinking.

File: .github/workflows/build_wheels_linux.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            package-name: torchdata
+    name: ${{ matrix.repository }}
+    uses: pytorch/test-infra/.github/workflows/build_wheels_linux.yml@main
+    with:
+      repository: ${{ matrix.repository }}
+      ref: ""
+      test-infra-repository: pytorch/test-infra
+      test-infra-ref: main
+      build-matrix: ${{ needs.generate-matrix.outputs.matrix }}
+      pre-script: ${{ matrix.pre-script }}
```
Comment (ejguan) [inline .github/workflows/build_wheels_linux.yml:38]: I think we need to do an extra step in `pre-script` for linux wheels. This is needed to convert torchdata wheels to `manylinux` compatible. I just checked the 3.7 release, it's not `manylinux2014` compatible by `auditwheel show torchdata-0.6.0.dev20221216+cpu-cp37-cp37m-linux_x86_64.whl`. And, apparently, `auditwheel repair` won't help.

Reason behind it is our C-Extension relying on CMake system from AWSSDK, which requires OpenSSL. I am not sure my approach is the best to make it manylinux compatible. Just want to share what I have done, I basically install static OpenSSL/libcrypto library manually.
https://github.com/pytorch/data/blob/0a0ae5d35be439c786c9ab798ea1596238d8640a/.github/workflows/_build_test_upload.yml#L124-L131 and specify those paths to let AWSSDK statically linking against OpenSSL https://github.com/pytorch/data/blob/0a0ae5d35be439c786c9ab798ea1596238d8640a/.github/workflows/_build_test_upload.yml#L153-L156

**Review by ejguan — APPROVED** [review]
Thank you for this effort. This is super helpful. Considering we might need to spend way more time on the manylinux compatible support, I am not going to block this PR since the wheels are not going to be uploaded. We can have a follow up PR to support it.

File: .github/workflows/build_wheels_linux.yml (hunk 3 above)
Comment (ejguan) [reply .github/workflows/build_wheels_linux.yml:38]: It should be similar to what TorchAudio is doing in https://github.com/pytorch/audio/blob/main/packaging/pre_build_script.sh

**Review by NivekT — APPROVED** [review]
Thank you so much! I will re-import and land after the last issue is resolved.

File: .github/workflows/build_wheels_linux.yml (hunk 3 above)
Comment (DanilBaibak) [reply .github/workflows/build_wheels_linux.yml:38]: We have the `pre-script` option exactly for this purpose. I agree, let's merge this PR and then start working on follow up PR to support manylinux compatible.

**Pull request closed without merging**